### DISCUSSION TITLE
Analytics: Instrument Azure Application Insights

### DISCUSSION THREAD

**Discussion by sunker** [grafana/grafana#38545] [General]
**What I'd like to do**
Add support for instrumenting Application Insights with events coming from the Grafana frontend. This should be an opt-in feature that can be configured in grafana.ini. 

**What is Application Insights?**
Azure Application Insights an Application Performance Management system that can be used to monitor live applications. By installing an instrumentation script on the web client, it allows you to ingest application / analytics data into the Application Insights data base. 

**How?**
We should use the [EchoSrv](https://github.com/grafana/grafana/blob/main/packages/grafana-runtime/src/services/EchoSrv.ts) to collect frontend events and create a new backend implementation for Application Insights, similar to how this is already done for Google Analytics and Rudderstack. 

An instrumentation key is required to initialize the Application Insights instrumentation script. This should be a configured in [grafana.ini](https://grafana.com/docs/grafana/latest/administration/configuration/#analytics). 

The Application Insights api has the following actions:
```javascript
appInsights.trackEvent({name: 'some event'});
appInsights.trackPageView({name: 'some page'});
appInsights.trackPageViewPerformance({name : 'some page', url: 'some url'});
appInsights.trackException({exception: new Error('some error')});
appInsights.trackTrace({message: 'some trace'});
appInsights.trackMetric({name: 'some metric', average: 42});
appInsights.trackDependencyData({absoluteUrl: 'some url', responseCode: 200, method: 'GET', id: 'some id'});
appInsights.startTrackPage("pageName");
appInsights.stopTrackPage("pageName", null, {customProp1: "some value"});
appInsights.startTrackEvent("event");
appInsights.stopTrackEvent("event", null, {customProp1: "some value"});
```

Currently, the EchoSrv collects page view events (would map to  `appInsights.trackPageView({name: 'some page'});`) and interaction events (would map to `appInsights.trackEvent({name: 'some event'});`). I doubt Application Insights would need to track other types of events than that, but if so we'll have to add new mechanisms for that. Tracking exceptions for example, shouldn't be so hard to do. 

Feedback appreciated. cc @daniellee @dprokop

**Comment by daniellee** [comment]
@sunker LGTM - looks like the interfaces for the EchoSrv and AppInsights line up nicely.

**Comment by daniellee** [comment]
# Next phase of the User Analytics Integraion for Application Insights - Frontend Error Tracking

The initial integration is simple and there is a lot more we could. The main goal with phase 2 of the integration is to track if users are having problems with data sources on a Grafana instance. 

## Error tracking on the frontend

This is the most obvious step as we already have an implementation for Sentry that logs frontend errors. This would catch bugs in frontend Grafana as well as any API calls that resulted in errors.

### Current implementation of error tracking in Grafana

Grafana has a Sentry integration that is used in Grafana Cloud for error tracking (see Frontend Logging PR: https://github.com/grafana/grafana/pull/28073). This is a custom integration that would have to be adapted for Application Insights but it has almost everything needed already. 

It is an opt-in integration that can either send events to Sentry, or to the Grafana backend to be logged to stdout. Saving to the Grafana backend will save the frontend errors in the Grafana server log which will save it in a log file or database like Loki if so configured. It even adds sourcemaps to stack traces (https://github.com/grafana/grafana/pull/30590) to make them more useful.

### Application Insights and error tracking

The hook for logging frontend exceptions is in: https://github.com/grafana/grafana/blob/main/packages/grafana-runtime/src/utils/logging.ts#L49. 

It currently calls out to the captureException function which is a Sentry function. We need to add a call to Application Insights exception tracking:

```
appInsights.trackException({exception: new Error('some error')});
```
### Needs more investigation

- We could just hardcode this in but should investigate doing this in a more generic way then the current Sentry-specific code.
- We also need to investigate if we need to add sourcemaps to the stack traces in the reported errors. We do this on the backend for Sentry currently and not in the frontend code.

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
